cmds/boot/boot/boot.go

```diff
@@ -79,171 +78,78 @@ func updateBootCmdline(cl string) string {
+	// Make changes to the kernel command line based on our cmdline.
+	if li, ok := img.(*boot.LinuxImage); ok {
+		li.Cmdline = updateBootCmdline(li.Cmdline)
+	}
 
-	// find or create the mountpoint.
-	m := filepath.Join(uroot, d)
-	if _, err = os.Stat(m); err != nil && os.IsNotExist(err) {
-		err = os.MkdirAll(m, 0777)
+	log.Printf("BootImage: %s", img)
+	if err := img.Load(*verbose); err != nil {
+		log.Printf("kexec load of %v failed: %v", img, err)
+		return
 	}
-	if err != nil {
-		debug("Can't make %v", m)
-		return err
+
+	if err := mp.Unmount(mount.MNT_DETACH); err != nil {
+		log.Printf("Can't unmount %v: %v", mp, err)
 	}
-	for _, filesystem := range supportedFilesystem {
-		var flags = uintptr(syscall.MS_RDONLY)
-		debug("\twith %v", filesystem)
-		if err := syscall.Mount(d, m, filesystem, flags, ""); err == nil {
-			return err
-		}
+	if *dryRun {
+		return
 	}
-	debug("No mount succeeded")
-	return fmt.Errorf("Unable to mount any partition on %v", d)
-}
 
-func umountEntry(n string) error {
-	return syscall.Unmount(n, syscall.MNT_DETACH)
+	if err := boot.Execute(); err != nil {
+		log.Printf("boot.Execute of %v failed: %v", img, err)
+	}
+	// TODO: We should probably return a real error.
+	return
 }
 
 // Localboot tries to boot from any local filesystem by parsing grub configuration
 func Localboot() error {
-	fs, err := getSupportedFilesystem()
-	if err != nil {
-		return errors.New("No filesystem support found")
-	}
-	debug("Supported filesystems: %v", fs)
-	sysList, err := filepath.Glob(*devGlob)
+	blockDevs, err := block.GetBlockDevices()
 	if err != nil {
-		return errors.New("No available block devices to boot from")
-	}
-	// The Linux /sys file system is a bit, er, awkward. You can't find
-	// the device special in there; just everything else.
-	var blkList []string
-	for _, b := range sysList {
-		blkList = append(blkList, filepath.Join("/dev", filepath.Base(b)))
+		return errors.New("no available block devices to boot from")
 	}
 
-	// We must validate if the MBR is bootable or not and keep the
-	// devices which do have such support drive are easy to
-	// detect.  This whole loop is pretty bogus at present, it
-	// assumes the first partiton we find with grub.cfg is the one
-	// we want. It works for now but ...
-	var allparts []string
-	for _, d := range blkList {
-		err := checkForBootableMBR(d)
-		if err != nil {
-			// Not sure it matters; there can be many bogus entries?
-			debug("MBR for %s failed: %v", d, err)
-			continue
-		}
-		debug("Bootable device %v found", d)
-		// You can't just look for numbers to match. Consider names like
-		// mmcblk0, where has parts like mmcblk0p1. Just glob.
-		g := d + "*"
-		all, err := filepath.Glob(g)
-		if err != nil {
-			log.Printf("Glob for all partitions of %s failed: %v", g, err)
-		}
-		allparts = append(allparts, all...)
-	}
-	uroot, err = ioutil.TempDir("", "u-root-boot")
+	// Try to only boot from "good" block devices.
+	blockDevs = blockDevs.FilterZeroSize()
```

Comment:
Would it make sense to sort these in some way so the order is consistent every time?

Reply:
I want to keep the order that the boot command had before, but this does bring up the problem of "I've messed up the boot order this command used to have" and I'm not sure what the correct boot order is. 

We should probably:

1) prefer entries from disks that have an MBR? (just as an indicator)
2) prefer entries from EFI system partitions?